src/main/java/org/apache/sysml/runtime/controlprogram/paramserv/LocalPSWorker.java

```diff
@@ -0,0 +1,157 @@
+import org.apache.sysml.parser.DMLProgram;
+import org.apache.sysml.parser.DataIdentifier;
+import org.apache.sysml.parser.Statement;
+import org.apache.sysml.runtime.DMLRuntimeException;
+import org.apache.sysml.runtime.controlprogram.FunctionProgramBlock;
+import org.apache.sysml.runtime.controlprogram.caching.MatrixObject;
+import org.apache.sysml.runtime.controlprogram.context.ExecutionContext;
+import org.apache.sysml.runtime.instructions.cp.CPOperand;
+import org.apache.sysml.runtime.instructions.cp.Data;
+import org.apache.sysml.runtime.instructions.cp.FunctionCallCPInstruction;
+import org.apache.sysml.runtime.instructions.cp.ListObject;
+
+public class LocalPSWorker extends PSWorker implements Runnable {
+
+	protected static final Log LOG = LogFactory.getLog(LocalPSWorker.class.getName());
+	private static final int POLL_FREQUENCY = 1;
+
+	private String _funcNS;
+	private String _funcName;
+	private ArrayList<DataIdentifier> _inputs;
+	private ArrayList<DataIdentifier> _outputs;
+	private CPOperand[] _boundInputs;
+	private ArrayList<String> _inputNames;
+	private ArrayList<String> _outputNames;
+
+	public LocalPSWorker(long workerID, String updFunc, Statement.PSFrequency freq, int epochs, long batchSize,
+			ListObject hyperParams, ExecutionContext ec, ParamServer ps) {
+		super(workerID, updFunc, freq, epochs, batchSize, hyperParams, ec, ps);
+
+		// Get the update function
+		String[] keys = DMLProgram.splitFunctionKey(_updFunc);
+		_funcName = keys[0];
+		if (keys.length == 2) {
+			_funcNS = keys[0];
+			_funcName = keys[1];
+		}
+		FunctionProgramBlock func = _ec.getProgram().getFunctionProgramBlock(_funcNS, _funcName);
+		_inputs = func.getInputParams();
+		_outputs = func.getOutputParams();
+		_boundInputs = _inputs.stream()
+				.map(input -> new CPOperand(input.getName(), input.getValueType(), input.getDataType()))
+				.toArray(CPOperand[]::new);
+		_inputNames = _inputs.stream().map(DataIdentifier::getName).collect(Collectors.toCollection(ArrayList::new));
+		_outputNames = _outputs.stream().map(DataIdentifier::getName).collect(Collectors.toCollection(ArrayList::new));
+	}
+
+	@Override
+	public void run() {
+
+		ParamservUtils.populate(_ec, _hyperParams, _inputs);
+
+		long dataSize = _features.getNumRows();
+
+		for (int i = 0; i < _epochs; i++) {
+			int totalIter = (int) Math.ceil(dataSize / _batchSize);
+			for (int j = 0; j < totalIter; j++) {
+				// Pull the global parameters from ps
+				ListObject globalParams = pollGlobalParams();
+				ParamservUtils.populate(_ec, globalParams, _inputs);
+
+				long begin = j * _batchSize + 1;
+				long end = begin + _batchSize;
+				if (j == totalIter - 1) {
+					// If it is the last iteration
+					end = dataSize;
+				}
+
+				// Get batch features and labels
+				MatrixObject bFeatures = ParamservUtils.sliceMatrix(_features, begin, end);
+				MatrixObject bLabels = ParamservUtils.sliceMatrix(_labels, begin, end);
+				_ec.setVariable(Statement.PS_FEATURES, bFeatures);
```

Comment:
Please do not compile custom DAGs per batch - either compile a parameterized instruction where you beed begin and end as scalar variables or use MatrixBlock.slice directly.